components/fxa-client/src/auth.rs

```diff
@@ -51,41 +42,47 @@ impl FirefoxAccount {
     ///   - `metrics` - optionally, additional metrics tracking parameters.
     ///       - These will be included as query parameters in the resulting URL.
     #[handle_error(Error)]
-    pub fn begin_oauth_flow<T: AsRef<str>>(
+    pub fn connect_with_oauth<T: AsRef<str>>(
         &self,
         // Allow both &[String] and &[&str] since UniFFI can't represent `&[&str]` yet,
         scopes: &[T],
         entrypoint: &str,
         metrics: Option<MetricsParams>,
-    ) -> ApiResult<String> {
+    ) -> ApiResult<()> {
         let scopes = scopes.iter().map(T::as_ref).collect::<Vec<_>>();
-        self.internal
-            .lock()
-            .begin_oauth_flow(&scopes, entrypoint, metrics)
+
+        // Lock self.internal to determine the oauth URL
+        let mut internal = self.internal.lock();
+        let (url, flow) = internal.begin_oauth_flow(&scopes, entrypoint, metrics)?;
+
+        // Release the lock while blocking on the OAuth handler
+        drop(internal);
```

Comment:
nit: What do we think about introducing another scope (i.e a `{ ` block) that way we avoid the explicit drop and make easier for the reader that the lock only lives within that scope